crates/pathfinder/src/state/ethereum_mockall.rs

```diff
@@ -0,0 +1,635 @@
+struct ChannelClosedError;
+
+impl EventSender {
+    /// Sends [Event::GetUpdate] on its channel and returns the result.
+    async fn get_update(
+        &self,
+        block: StarknetBlockNumber,
+    ) -> Result<Option<StateUpdateLog>, ChannelClosedError> {
+        let (tx, rx) = oneshot::channel();
+        self.0
+            .send(Event::GetUpdate(block, tx))
+            .await
+            .map_err(|_send_err| ChannelClosedError)?;
+
+        rx.await.map_err(|_recv_err| ChannelClosedError)
+    }
+
+    /// Sends [Event::Updates] on its channel.
+    async fn updates(&self, updates: Vec<StateUpdateLog>) -> Result<(), ChannelClosedError> {
+        self.0
+            .send(Event::Updates(updates))
+            .await
+            .map_err(|_send_err| ChannelClosedError)
+    }
+
+    /// Sends [Event::Reorg] on its channel.
+    async fn reorg(&self, block: StarknetBlockNumber) -> Result<(), ChannelClosedError> {
+        self.0
+            .send(Event::Reorg(block))
+            .await
+            .map_err(|_send_err| ChannelClosedError)
+    }
+}
+
+async fn sync_ethereum_state_impl(
+    mut eth_api: impl EthereumApi,
+    tx_event: mpsc::Sender<Event>,
+) -> anyhow::Result<()> {
+    let event_sender = EventSender(tx_event);
+    loop {
+        match eth_api.fetch_logs().await {
+            Ok(logs) => {
+                // If empty, then we are at head of chain, sleep a bit and try again.
+                if logs.is_empty() {
+                    tokio::time::sleep(Duration::from_secs(5)).await;
+                    continue;
+                }
+
+                // There were log updates, send the event!
+                if let Err(_exit) = event_sender.updates(logs).await {
+                    return Ok(());
+                }
+            }
+            Err(FetchError::Reorg) => {
+                // Unwrap is safe as it is not be possible to get a reorg event if there
+                // was no latest log to reorg against. We know that this block already needs to
+                // be reorg'd since it triggered the reorg in the first place.
+                let mut reorg_tail = eth_api.log_head().clone().unwrap();
+
+                // Check each Starknet block in reverse history order, until we find a still
+                // valid block. This becomes the new head of our L1 state.
+                let new_head = loop {
+                    // We have reached Starknet genesis, no older blocks to check.
+                    if reorg_tail.block_number == StarknetBlockNumber::GENESIS {
+                        break None;
+                    }
+
+                    // Reqeuest the previous Starknet block update.
+                    let update = match event_sender
+                        .get_update(reorg_tail.block_number.decrement())
+                        .await
+                    {
+                        Ok(update) => update,
+                        Err(_exit) => return Ok(()),
+                    };
```

Comment:
I can imagine this "event" (which is really a request for a database read), should also come from a separate database API instead of being part of this event queue.

This would also neaten things up a lot I think.